catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/security/accesscontrol/AccessControlPreQueryPlugin.java

```diff
@@ -0,0 +1,237 @@
+import ddf.catalog.plugin.StopProcessingException;
+import ddf.security.Subject;
+import ddf.security.SubjectIdentity;
+import ddf.security.SubjectUtils;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Collectors;
+import org.apache.commons.collections.CollectionUtils;
+import org.apache.shiro.SecurityUtils;
+import org.geotools.filter.visitor.SimplifyingFilterVisitor;
+import org.opengis.filter.Filter;
+import org.osgi.framework.ServiceReference;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Plugin that looks for queries that target access controlled metacards (workspaces, forms, etc)
+ * and updates the query filter so part of the access control security policy is baked in, removing
+ * a significant burden from post-query filtering.
+ *
+ * <p><b>This plugin does not add any new policy. It only optimizes existing policy by adding the
+ * rules to the {@link Filter} so the database does some heavy lifting for us.</b>
+ *
+ * <p>When identifying a {@link Filter} that only selects access controlled metacards, the current
+ * implementation takes a best effort approach on the entire filter as a whole. It does not perform
+ * modification on individual subtrees, only the entire filter, or not at all.
+ *
+ * <p>This plugin will skip processing queries when:
+ *
+ * <ul>
+ *   <li>The user has the {@link AccessControlSecurityConfiguration} group.
+ *   <li>When no predicates operating on {@link Core#METACARD_TAGS} were found in the filter.
+ *   <li>There was at least one tags predicate on the filter that was not in the {@link
+ *       #accessControlledTags} set.
+ * </ul>
+ *
+ * This plugin will ignore {@link Core#METACARD_TAGS} present on the filter if:
+ *
+ * <ul>
+ *   <li>The predicate is a negative check, such as {@link org.opengis.filter.PropertyIsNotEqualTo}.
+ *   <li>The predicate falls under the logical composite {@link org.opengis.filter.Not} operator.
+ * </ul>
+ */
+public class AccessControlPreQueryPlugin implements PreQueryPlugin {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(AccessControlPreQueryPlugin.class);
+
+  private static final String ACCESS_CONTROLLED_TAG = "access-controlled-tag";
+
+  private final Set<String> accessControlledTags = new HashSet<>();
+
+  private final FilterBuilder filterBuilder;
+
+  private final SubjectIdentity identity;
+
+  private final AccessControlSecurityConfiguration configuration;
+
+  public AccessControlPreQueryPlugin(
+      FilterBuilder filterBuilder,
+      SubjectIdentity identity,
+      AccessControlSecurityConfiguration configuration) {
+    this.filterBuilder = filterBuilder;
+    this.identity = identity;
+    this.configuration = configuration;
+  }
+
+  // Verify using this class in a reference listener is sufficient
+  private static Set<String> aggregateTags(List<ServiceReference<?>> references) {
```

Comment:
my IDE marks this as unused. if it is used, consider adding a suppresswarnings unused w/ an explanation

Reply:
Code tied to the above comment. It was unused and has been removed.